_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,753 @@
+
+num_epochs <- 100
+
+for (epoch in 1:num_epochs) {
+  
+  model$train()
+  batch_losses <- c()
+  
+  for (b in enumerate(dl)) {
+    
+    optimizer$zero_grad()
+    
+    # last-time-step output from last layer
+    preds <- model(b$x)[[2]][[2]][[1]]
+  
+    loss <- nnf_mse_loss(preds, b$y)
+    batch_losses <- c(batch_losses, loss$item())
+    
+    loss$backward()
+    optimizer$step()
+  }
+  
+  if (epoch %% 10 == 0)
+    cat(sprintf("\nEpoch %d, training loss:%3f\n", epoch, mean(batch_losses)))
+}
+
+```
+
+    Epoch 10, training loss:0.008522
+
+    Epoch 20, training loss:0.008079
+
+    Epoch 30, training loss:0.006187
+
+    Epoch 40, training loss:0.003828
+
+    Epoch 50, training loss:0.002322
+
+    Epoch 60, training loss:0.001594
+
+    Epoch 70, training loss:0.001376
+
+    Epoch 80, training loss:0.001258
+
+    Epoch 90, training loss:0.001218
+
+    Epoch 100, training loss:0.001171
+
+Loss decreases, but that in itself is not a guarantee the model has learned anything. Has it? Let's inspect its forecast for the very first sequence and see.
+
+For printing, I'm zooming in on the relevant region in the 24x24-pixel frame. Here is the ground truth for time step six:
+
+```{r}
+b$y[1, 1, 6:15, 10:19]
+```
+
+    0  0  0  0  0  0  0  0  0  0
+    0  0  0  0  0  0  0  0  0  0
+    0  0  1  0  0  0  0  0  0  0
+    0  0  0  1  0  0  0  0  0  0
+    0  0  0  0  1  0  0  0  0  0
+    0  0  0  0  0  1  0  0  0  0
+    0  0  0  0  0  0  1  0  0  0
+    0  0  0  0  0  0  0  1  0  0
+    0  0  0  0  0  0  0  0  0  0
+    0  0  0  0  0  0  0  0  0  0
+
+And here is the forecast. This does not look bad at all, given there was neither experimentation nor tuning involved.
+
+```{r}
+round(as.matrix(preds[1, 1, 6:15, 10:19]), 2)
+```
+
+           [,1]  [,2]  [,3]  [,4]  [,5]  [,6]  [,7]  [,8]  [,9] [,10]
+     [1,]  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00     0
+     [2,] -0.02  0.36  0.01  0.06  0.00  0.00  0.00  0.00  0.00     0
+     [3,]  0.00 -0.01  0.71  0.01  0.06  0.00  0.00  0.00  0.00     0
+     [4,] -0.01  0.04  0.00  0.75  0.01  0.06  0.00  0.00  0.00     0
+     [5,]  0.00 -0.01 -0.01 -0.01  0.75  0.01  0.06  0.00  0.00     0
+     [6,]  0.00  0.01  0.00 -0.07 -0.01  0.75  0.01  0.06  0.00     0
+     [7,]  0.00  0.01 -0.01 -0.01 -0.07 -0.01  0.75  0.01  0.06     0
+     [8,]  0.00  0.00  0.01  0.00  0.00 -0.01  0.00  0.71  0.00     0
+     [9,]  0.00  0.00  0.00  0.01  0.01  0.00  0.03 -0.01  0.37     0
+    [10,]  0.00  0.00  0.00  0.00  0.00  0.00 -0.01 -0.01 -0.01     0
+
+This should suffice for a sanity check. If you made it till the end, thanks for your patience! In the best case, you'll be able to apply this architecture (or a similar one) to your own data -- but even if not, I hope you've enjoyed learning about `torch` model coding and/or RNN weirdness ;-)
+
+Myself, I am certainly looking forward to exploring convLSTMs on real-world problems in the near future. Thanks for reading!
```

Comment:
Myself is a reflexive pronoun, which you only use when it's the object of the sentence. The best way I can think of it is with french verbs, _se laver_ to wash (you wash yourself). You can put yourself to bed, bathe yourself, kill yourself…there aren't a lot of reflexive verbs in English (or at least I can't think of them right now). Basically, people use myself incorrectly all the time (including above). Myself can also be used as an intensive pronoun, but, in that case, it should go after the "I"— _I myself am certainly looking forward to…_